pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -184,18 +188,36 @@ func (e *EndpointController) Run(workers int, stopCh <-chan struct{}) {
 
 func (e *EndpointController) getPodServiceMemberships(pod *v1.Pod) (sets.String, error) {
 	set := sets.String{}
-	services, err := e.serviceLister.GetPodServices(pod)
+	allServices, err := e.serviceLister.Services(pod.Namespace).List(labels.Everything())
 	if err != nil {
 		// don't log this error because this function makes pointless
 		// errors when no services match.
 		return set, nil
 	}
-	for i := range services {
-		key, err := controller.KeyFunc(services[i])
-		if err != nil {
-			return nil, err
+
+	var selector labels.Selector
+	for _, service := range allServices {
+		cachedSelector, ok := e.serviceSelectorCache.Load(service)
+		if ok {
+			// retrieve selector from cache
+			selector = cachedSelector.(labels.Selector)
+		} else {
+			// only create a selector if we don't have it cached already
+			if service.Spec.Selector == nil {
```

Comment:
The code seems to have come from serviceLister#GetPodServices.
What if serviceLister#GetPodServices gets changed in the future ? That may introduce discrepancy.